### PR TITLE
[SPARK-29874][SQL]Optimize Dataset.isEmpty()

### PR DESCRIPTION
**Pull request by AngersZhuuuu**
### What changes were proposed in this pull request?
In  origin way to judge if a DataSet is empty by 
```
 def isEmpty: Boolean = withAction("isEmpty", limit(1).groupBy().count().queryExecution) { plan =>
    plan.executeCollect().head.getLong(0) == 0
  }
```
will add two shuffles by `limit()`, `groupby() and count()`, then collect all data to driver.
In this way we can avoid `oom` when collect data to driver. But it will trigger all partitions calculated and add more shuffle process.

We change it to 
```
  def isEmpty: Boolean = withAction("isEmpty", select().queryExecution) { plan =>
    plan.executeTake(1).isEmpty
  }
```
After these pr, we will add a column pruning to origin LogicalPlan and use `executeTake()` API.
then we won't add more shuffle process and just compute only one partition's data in last stage.
In this way we can reduce cost when we call `DataSet.isEmpty()` and won't bring memory issue to driver side.

### Why are the changes needed?
Optimize Dataset.isEmpty()

### Does this PR introduce any user-facing change?
No

### How was this patch tested?
Origin UT

### CODE REVIEW

**Comment by AngersZhuuuu** [comment]
cc @cloud-fan

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala [hunk 1]
```diff
@@ -586,8 +586,8 @@ class Dataset[T] private[sql](
    * @group basic
    * @since 2.4.0
    */
-  def isEmpty: Boolean = withAction("isEmpty", limit(1).groupBy().count().queryExecution) { plan =>
-    plan.executeCollect().head.getLong(0) == 0
+  def isEmpty: Boolean = withAction("isEmpty", queryExecution) { plan =>
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: shall we do column pruning?

**Comment by cloud-fan** [comment]
Can we have some benchmark numbers?

**Comment by cloud-fan** [comment]
ok to test

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala (hunk 1 above)
Comment (AngersZhuuuu) [reply sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: > shall we do column pruning?

Of course, add it.

**Comment by maropu** [comment]
Can you address more in the PR desc for better commit logs, how to optimize it?

**Comment by srowen** [comment]
Ping @AngersZhuuuu

**Comment by AngersZhuuuu** [comment]
> Ping @AngersZhuuuu

Thank you ping, sorry for pending this work.  A little busy these days. 
Starting work on these things.

**Comment by AngersZhuuuu** [comment]
```
  test("benchmark of empty") {
    var start = System.currentTimeMillis()
    var isEmpty = spark.range(10000000)
      .repartition(100)
      .limit(1)
      .groupBy()
      .count()
      .queryExecution.executedPlan.executeCollect().head.getLong(0) == 0
    println(isEmpty)
    var end = System.currentTimeMillis()
    // scalastyle:off
    println(s"duration = ${end - start}")

    start = System.currentTimeMillis()
    isEmpty = spark.range(10000000)
      .repartition(100)
      .select()
      .queryExecution.executedPlan.executeTake(1) == 0
    println(isEmpty)
    end = System.currentTimeMillis()
    // scalastyle:off
    println(s"duration = ${end - start}")
  }

Result
false
duration = 7248
false
duration = 1449
```

@cloud-fan @maropu @srowen   
The test case is simple but  can mimic the behavior before and after the API change.

**Comment by cloud-fan** [comment]
great! can you enrich the PR description? `Optimize Dataset.isEmpty()` is good in the "Why" section but we need to put more in the "What" section. e.g. we change the implementation to avoid shuffles.

**Comment by AngersZhuuuu** [comment]
> great! can you enrich the PR description? `Optimize Dataset.isEmpty()` is good in the "Why" section but we need to put more in the "What" section. e.g. we change the implementation to avoid shuffles.

Updated , is clear now?

**Comment by cloud-fan** [comment]
> will add three shuffles by limit(), groupby() and count()

have you confirmed? groupby + count is one operator called Aggregate.

**Comment by AngersZhuuuu** [comment]
> > will add three shuffles by limit(), groupby() and count()
> 
> have you confirmed? groupby + count is one operator called Aggregate.

Updated,  `count` won't trigger shuffle.

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Review by HyukjinKwon — COMMENTED** [review]
LGTM too

**Pull request closed without merging**